contracts/staking/StakingMain.sol

```diff
@@ -0,0 +1,179 @@
+        if (stakes.length != 0) {
+            Stake memory lastStake = stakes[stakes.length - 1];
+            if (lastStake.startDate > block.timestamp) revert StakingErrors.StakeDateBeforeNow();
+
+            uint256 timeDelta = block.timestamp - lastStake.startDate;
+            if (timeDelta < StakingConstants.STAKE_COOLDOWN) revert StakingErrors.TooOftenStaking();
+        }
+
+        // check staker balance
+        uint256 votesBalance = IERC20(_config.voteAddress).balanceOf(msg.sender);
+        if (votesBalance < votesNumber) revert StakingErrors.NotEnoughVotes();
+
+        // append a new stake
+        _pushStake(msg.sender, block.timestamp, votesNumber);
+
+        // TODO: change IERC20 to IVote
+        IERC20(_config.voteAddress).transferFrom(msg.sender, address(this), votesNumber);
+        emit VotesStaked(block.timestamp, votesNumber);
+    }
+
+    function unstakeVotes() external override {
+        Stake[] storage stakes = _stakes[msg.sender];
+        if (stakes.length == 0) revert StakingErrors.StakerNotFound();
+
+        // remove stakes
+        uint256 vouchersNumber;
+        uint256 votesNumber;
+        while (stakes.length > 0) {
+            Stake memory stake = stakes[stakes.length - 1];
+            vouchersNumber += _calculateVouchers(stake);
+            votesNumber += stake.amount;
+
+            _popStake(msg.sender);
+        }
+
+        // ???
+        uint256 votesBalance = IERC20(_config.voteAddress).balanceOf(address(this));
+        if (votesBalance < votesNumber) revert StakingErrors.NotEnoughVotes();
+
+        IERC20(_config.voteAddress).transfer(msg.sender, votesNumber);
+        emit VotesUnstaked(msg.sender, votesNumber);
+
+        if (vouchersNumber == 0) return;
+
+        IVoucher(_config.voucherAddress).mint(msg.sender, vouchersNumber);
+        emit VouchersMinted(msg.sender, vouchersNumber);
+    }
+
+    function withdrawVouchersForAll() external override isOwner {
+        for (uint256 i = 0; i < _stakers.length; i++) {
+            withdrawVouchers(_stakers[i]);
+        }
+    }
+
+    function setThreshold(uint256 threshold) external override isOwner {
+        _votesThreshold = threshold;
+    }
+
+    function getThreshold() external view override returns (uint256) {
+        return _votesThreshold;
+    }
+
+    function getVotesAmount() external view override returns (uint256) {
+        Stake[] memory stakes = _stakes[msg.sender];
+        if (stakes.length == 0) revert StakingErrors.StakerNotFound();
+
+        uint256 voteNumber;
+        for (uint256 i = 0; i < stakes.length; i++) {
+            voteNumber += stakes[i].amount;
+        }
+        return voteNumber;
+    }
+
+    function getVouchersAmount() external view override returns (uint256) {
+        Stake[] memory stakes = _stakes[msg.sender];
+        if (stakes.length == 0) revert StakingErrors.StakerNotFound();
+
+        uint256 vouchersNumber;
+        for (uint256 i = 0; i < stakes.length; i++) {
```

Comment:
Put length in front of the loop "uint length = _stakers.length;" and use it as a higher bound in it. Better for gas consumption